swap/cheque.go

```diff
@@ -0,0 +1,119 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// Encode encodes the cheque in the format used in the signing procedure
+func (cheque *Cheque) Encode() []byte {
+	serialBytes := make([]byte, 32)
+	amountBytes := make([]byte, 32)
+	timeoutBytes := make([]byte, 32)
+	// we need to write the last 8 bytes as we write a uint64 into a 32-byte array
+	// encoded in BigEndian because EVM uses BigEndian encoding
+	binary.BigEndian.PutUint64(serialBytes[24:], cheque.Serial)
+	binary.BigEndian.PutUint64(amountBytes[24:], cheque.Amount)
+	binary.BigEndian.PutUint64(timeoutBytes[24:], cheque.Timeout)
+	// construct the actual cheque
+	input := cheque.Contract.Bytes()
+	input = append(input, cheque.Beneficiary.Bytes()...)
+	input = append(input, serialBytes[:]...)
+	input = append(input, amountBytes[:]...)
+	input = append(input, timeoutBytes[:]...)
+
+	return input
+}
+
+// sigHash hashes the cheque using the prefix that would be added by eth_Sign
+func (cheque *Cheque) sigHash() []byte {
+	input := crypto.Keccak256(cheque.Encode())
+	withPrefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(input), input)
+	return crypto.Keccak256([]byte(withPrefix))
+}
+
+// VerifySig verifies the signature on the cheque
+func (cheque *Cheque) VerifySig(expectedSigner common.Address) error {
+	sigHash := cheque.sigHash()
+
+	if cheque.Sig == nil {
+		return fmt.Errorf("tried to verify signature on cheque with sig nil")
+	}
+	// copy signature to avoid modifying the original
+	sig := make([]byte, len(cheque.Sig))
+	copy(sig, cheque.Sig)
+	// reduce the v value of the signature by 27 (see signContent)
+	sig[len(sig)-1] -= 27
+	pubKey, err := crypto.SigToPub(sigHash, sig)
+	if err != nil {
+		return err
+	}
+
+	if crypto.PubkeyToAddress(*pubKey) != expectedSigner {
+		return ErrInvalidChequeSignature
+	}
+
+	return nil
+}
+
+// Sign signs the cheque with supplied private key
+func (cheque *Cheque) Sign(prv *ecdsa.PrivateKey) ([]byte, error) {
+	sig, err := crypto.Sign(cheque.sigHash(), prv)
+	if err != nil {
+		return nil, err
+	}
+	// increase the v value by 27 as crypto.Sign produces 0 or 1 but the contract only accepts 27 or 28
+	// this is to prevent malleable signatures. while not strictly necessary in this case the ECDSA implementation from Openzeppelin expects it.
+	sig[len(sig)-1] += 27
```

Comment:
When writing a test (it was obviously a programming error on my side) this line produced a `index out of range` panic, because `len(sig)` was 0. Maybe we should have a check for that to avoid that error

Reply:
Are you sure it was in the signing function? The only point where it should have been possible was the `sig[len(sig)-1] -= 27` in the verify function (where it no longer can happen because I introduced an `if cheque.Sig == nil` check in the last PR).

Here we can only reach this code if `crypto.Sign` was a success, at which point sig has to be non-empty.

Reply:
No I am not sure, I just remember it happened on such a `[len(sig) -1]` instruction. If you are 100% it can not happen, disregard, nevertheless don't forget that in tests, any of these functions can be called directly in the package, and for some reason someone might want to call this one directly (maybe, maybe not :) )

Reply:
Even if called directly in the test it should be impossible to get an out of range here (assuming no bugs in the `crypto` library).